### ISSUE TITLE
Unclear error when dynamic block iterator is wrong

### ISSUE THREAD

**Issue by RLRabinowitz** [opentofu/opentofu#1372]
### OpenTofu Version

```shell
1.6.2
```

### OpenTofu Configuration Files

```hcl
locals {
  actions = ["ec2:*", "s3:*"]
}

provider "aws" {
  region = "us-east-1"
}

data "aws_iam_policy_document" "doc" {
  dynamic "statement" {
    for_each = local.actions
    iterator = "client"

    content {
      actions = [client.value]
      resources = ["*"]
    }
  }
}

output "my-data" {
  value = data.aws_iam_policy_document.doc
}
```

### Debug Output

.

### Expected Behavior

Error saying that the iterator is wrong (it should be `client`).
In some other cases (like not using a `local` value in the `for_each`, you get a better error (albeit not the best):

```
│ Error: Invalid expression
│ 
│   on main.tf line 8, in data "aws_iam_policy_document" "doc":
│    8:     iterator = "client"
│ 
│ A single static variable reference is required: only attribute access and indexing with constant keys. No calculations, function calls, template expressions, etc are allowed here.
```

### Actual Behavior

```
│ Error: Unknown variable
│ 
│   on main.tf line 14, in data "aws_iam_policy_document" "doc":
│   14:     for_each = local.actions
│ 
│ There is no variable named "local".
```

### Steps to Reproduce

1. `tofu init`
2. `tofu plan`

### References

Related issue in the Terraform repository - https://github.com/hashicorp/terraform/issues/34132

**Comment by namgyalangmo** [comment]
I would like to resolve this issue!

**Comment by bunniseng** [comment]
Hi @RLRabinowitz 

I would like to resolve this issue, can you please assign it to me.

Thanks

**Comment by cam72cam** [comment]
Thanks @bunniseng, please double check that this is still an issue as we have updated HCL since this issue was created.

**Comment by bunniseng** [comment]
I can confirm the issue has been resolved for v1.7.0

Using the example provided by RLRabinowitz, the output is as follows:

```
╷
│ Error: Invalid expression
│ 
│   on main.tf line 12, in data "aws_iam_policy_document" "doc":
│   12:     iterator = "client"
│ 
│ A single static variable reference is required: only attribute access and indexing with constant keys. No calculations, function calls, template expressions, etc are allowed here.
╵
```

@cam72cam I think we can close this issue as resolved.

**Comment by cam72cam** [comment]
Thanks for confirming @bunniseng!
